File: 2024/Python/1/solution.py
import os

def solution():
    script_dir = os.path.dirname(os.path.realpath(__file__))
    input_file_path = os.path.join(script_dir, 'input.txt')

    with open(input_file_path) as file:
        data = file.read().splitlines()

    list1 = []
    list2 = []

    for i in range(len(data)):
        item1, item2 = data[i].split()

        list1.append(int(item1))
        list2.append(int(item2))

    list1.sort()
    list2.sort()

    print("list1:", list1)
    print("list2:", list2)

    totalDistance = 0
    totalSimilarityScore = 0

    for i in range(len(list1)):
        item1 = list1[i]

        totalDistance += abs(item1 - list2[i])

        amountOfOccurrencesInList2 = list2.count(item1)
        totalSimilarityScore += item1 * amountOfOccurrencesInList2

    print("total distance:", totalDistance)
    print("total similarity score:", totalSimilarityScore)

File: 2024/Go/1/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func count[T any](slice []T, f func(T) bool) int {
	count := 0
	for _, s := range slice {
		if f(s) {
			count++
		}
	}
	return count
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	r := bufio.NewReader(file)

	list1 := make([]int, 0)
	list2 := make([]int, 0)

	for {
		line, _, err := r.ReadLine()
		if len(line) > 0 {
			lineString := string(line)
			items := slices.DeleteFunc(strings.Split(lineString, " "), func(e string) bool { return e == "" })

			item1, _ := strconv.Atoi(items[0])
			item2, _ := strconv.Atoi(items[1])

			list1 = append(list1, item1)
			list2 = append(list2, item2)
		}
		if err != nil {
			break
		}
	}

	sortFunc := func(a, b int) int {
		return a - b
	}

	slices.SortFunc(list1, sortFunc)
	slices.SortFunc(list2, sortFunc)

	fmt.Printf("List1: %o\n\n", list1)
	fmt.Printf("List2: %o\n\n", list2)

	totalDistance := 0.0
	totalSimilarityScore := 0

	for i := 0; i < len(list1); i++ {
		item1 := list1[i]
		item2 := list2[i]
		totalDistance += math.Abs(float64(item1 - item2))

		amountOfOccurencesInList2 := count(list2, func(e int) bool { return e == item1 })
		totalSimilarityScore += item1 * amountOfOccurencesInList2
	}

	fmt.Printf("Total distance: %s\n", strconv.FormatFloat(totalDistance, 'f', -1, 64))
	fmt.Printf("Total similarity score: %d\n", totalSimilarityScore)
}
